winch/codegen/src/isa/x64/asm.rs

```diff
@@ -471,6 +472,90 @@ impl Assembler {
         });
     }
 
+    /// Vector load and extend.
+    pub fn xmm_pmov_mr(
+        &mut self,
+        src: &Address,
+        dst: WritableReg,
+        ext: VectorExtendKind,
+        flags: MemFlags,
+    ) {
+        assert!(dst.to_reg().is_float());
+
+        let op = match ext {
+            VectorExtendKind::V128Extend8x8S => SseOpcode::Pmovsxbw,
+            VectorExtendKind::V128Extend8x8U => SseOpcode::Pmovzxbw,
+            VectorExtendKind::V128Extend16x4S => SseOpcode::Pmovsxwd,
+            VectorExtendKind::V128Extend16x4U => SseOpcode::Pmovzxwd,
+            VectorExtendKind::V128Extend32x2S => SseOpcode::Pmovsxdq,
+            VectorExtendKind::V128Extend32x2U => SseOpcode::Pmovzxdq,
+        };
```

Comment:
Given that we're starting with avx2 as our baseline, I wonder if we should emit the `vpmov*` variants here.

Reply:
We could. I figured if they are equivalent instructions that are supported on older chipsets, we could emit those so there would be less work to support non-AVX CPUs in the future given we only need to support 128-bit vectors. One thing I'm not clear on is if `Vpmov*` instructions are faster or not. If they are faster, then it would probably make sense to emit those instead.

Reply:
My expectation is that they are probably more performant, but I haven't performed any meaningful benchmarks. Aside from the potential performance benefit, my comment here is for consistency mostly: given that we aren't planning on doing any checks/emission for anything other than avx2 for SIMD at least initially, I'd prefer if we could default to emitting avx2 only to start. My expectation is that when thinking about fallbacks for anything < avx2, it will require substantial work regardless given that we'd probably want to find a way to share that implementation with what already exists in Cranelift.

Reply:
It's been a bit since I last checked, and I also haven't benchmarked, but IIRC the Intel optimization guide recommended either using exclusively VEX instructions (ones with the `v*` prefix) or exclusively non-VEX instructions. It's got something subtle to do with CPU state and such. It generally has to do with the fact that VEX instructions mutate the entire register contents as opposed to SSE which only mutate the lower 128-bits. This means that SSE instructions can accidentally create false dependencies between instructions.

The rule of thumb I used in Cranelift at least was to try to unconditionally prefer `v*` VEX instructions when AVX is enabled as I think that's what Intel assumes compilers are doing to make things faster